### ISSUE TITLE
Question: What is the correct date format for additional fields?

### ISSUE THREAD

**Issue by sandreas** [Zeugma440/atldotnet#134]
## The problem
I would like to add `PurchaseDate` / `purd` for `MP4` by additional fields. Since that is only a string contained dictionary, what format would be accepted as Date?

What about other Dates in other tagging formats? Are they all the same or is there a difference?

I saw `2022/06/30` and `2022-06-30` (ISO 8601), as well as `2022-06-30T12:34:56Z` in the wild, but I'm not convinced, that this is correct everywhere.

**Comment by Zeugma440** [comment]
The answer is : "it depends on what the target tagging format is".

For instance, regarding **Recording date** :
- The M4A/MP4 native format expects either a plain year or an entire date (e.g. 1997-06-20) in the same `©day` field
- The ID3v2.4 standard expects an ISO timestamp in `TDRC` ;
- The ID3v2.3 standard expects a combination of fields (year `TYER`, daymonth `TDAT` and time `TIME`, all optional);
- The RIFF WAV native LIST INFO format expects something similar to an ISO timestamp (`ICRD` field), except the "T" that usually separates date and time is a space character;

=> you'd need to adapt what you're writing to the specs of the target format, which is what ATL does under the hood for standard fields.

If I were you, as far as MP4 does, I'd go with the same convention as `©day`.

Hope this helps 😉

**Comment by sandreas** [comment]
> Hope this helps 😉

Yes, indeed. 


> you'd need to adapt what you're writing to the specs of the target format, which is what ATL does under the hood for standard fields.

Would it be possible / sensible to "normalize" the string format used in `AdditionalFields` to `2022-04-06T19:50:00Z` and make `atldotnet` convert it via

```c#
if(DateTime.TryParse(AdditionalFields[key], out var dateTime)) {
    var normalizedDateTime = dateTime.ToString("<tagspecific-format>");
}
```
That way the user would not have to deal with date format shenanigans in `AdditionalFields`, because `atldotnet` already knows, which format to use for specific tags and has the tools to reformat everything correctly?!

That would be awesome.

**Comment by Zeugma440** [comment]
Heh, I was sure you'd come up with such a suggestion. I could handle that the same way as popularity, and open the API through some public helper class. Lemme think this through...

**Comment by Zeugma440** [comment]
`DateTime.TryParse` is way too expensive to implement what you suggested.

I eventually went for a prefix :

```csharp
theTag.AdditionalFields = new Dictionary<string, string>
{
     { "TEST", MetaDataHolder.DATETIME_PREFIX + now.ToFileTime() }
};
theFile.UpdateTagInFile(theTag, tagType);
```

**Comment by sandreas** [comment]
> DateTime.TryParse is way too expensive to implement it the way you suggested.

I'm not sure I got this right... My suggestion was not to `DateTime.TryParse` EVERY item in `AdditionalFields`, but only for fieldnames known to be a `DateTime` value (like `purd` for `PurchaseDate`). The only thing required for this would be a specification of date fields in the format file:

```c#

class MP4 : MetaDataIO, IAudioDataIO
{
        private readonly string[] DateTimeFieldNames { get;  set; } = {"purd" /*,...*/}


        private int writeFrames(TagData tag, BinaryWriter w) {
        // ...
        foreach(var key in DateTimeFieldNames) {
              if(!tag.AdditionalFields.ContainsKey(key)) {
                    continue;
              }
              
              AdditionalFields[key].PreFormatHook = ConvertDateTime;
              // or
              AdditionalFields[key] = ConvertDateTime(AdditionalFields[key]);
        }


        }
}
```

> I eventually went for a prefix

I'm sorry I have to ask again: How would a prefix help with getting rid of date format problems? Maybe I did get this correctly...

**Comment by Zeugma440** [comment]
> I'm not sure I got this right... My suggestion was not to DateTime.TryParse EVERY item in AdditionalFields
My mistake, I thought that's what you had in mind

> I'm sorry I have to ask again: How would a prefix help with getting rid of date format problems?

No problem. Your suggestion requires ATL to publish some hook / public method for the client app to convert dates to the format the tagging system expects. It becomes complicated to handle when you're working on a track with multiple taggings requiring different formats (e.g. MP3 with both ID3v2.4 and APE tags).

With the solution I chose, it's ATL's responsibilty to format the date at the right moment, with the correct pattern. The client app is just telling ATL which `AdditionalField` value is a date by prefixing that value with a constant string. Whenever ATL wants to write `AdditionalField`s, if it finds the prefix, it formats the date properly.

Is that clearer for you ?

**Comment by sandreas** [comment]
> Is that clearer for you ?

Yes, thank you very much... So maybe I summarize what I understood, and what I can't wrap my head around:

- A file MAY store different metadata formats (e.g. `flac` supports at least `id3v2` and `APE`)
- Usually, a file DOES store one *main* metadata format (e.g. `mp3` mostly uses `id3v2.3`)
- Some files have a NATIVE format to store (like `mp4` with its `atoms`) and do not support multiple formats
- The metadata format and the file format may be completely independent (whille most of the time, they are not)
- The `Track` class tries to give access to metadata as simple as possible, by providing a lot of  *background magic*

So what I can't wrap my head around is:

- `Track.AudioFormat` does not help, if you try to set fields for a specific metadata format (it just has `mp3` but not `id3v2`)
- `Track.AdditionalFields` is not accurate enough, because it implements metadata format specific keys, but if you would like to support multiple, that is not possible via `KeyValuePair`

What I would have understood is:

- `Track.Metadata.AdditionalFields` for the default metadata format (e.g. `id3v2.3` or the native one)
- `Track.MetadataContainers[<formatkey>].AdditionalFields` for other formats, also including `id3v2.3`
- `Track.MetadataContainers[<formatkey>]` being a real metadata format class, not file format class (like `Id3v2`, `Mp4`, `Ape` etc.)

That way it would be more *understandable* to me :-) But lets dicuss this, after I took a deep dive into the code.

**Comment by Zeugma440** [comment]
Everything in your 1st paragraph is right. Just be careful of "usually" and "most of the time", as these are _not_ reasons for simplifying the library. If the specs say there may be multiple values, the library has to support that, no matter how little sample files actually have that specific setting. Devil's in the details 😉 

> Track.AudioFormat does not help, if you try to set fields for a specific metadata format (it just has mp3 but not id3v2)

Because that information belongs to the "audio" realm, and not the "metadata" realm. The whole topic that gave birth to that field is there : #78

> Track.AdditionalFields is not accurate enough, because it implements metadata format specific keys, but if you would like to support multiple, that is not possible via KeyValuePair

The primary purpose of `Track.AdditionalFields` is to provide a super easy and intuitive way to read and edit fields that are not supported through `Track` / `IMetaData` standardized properties - either because the field is too specific to be standardized (e.g. Audible's "Amazon catalog number" field) or because it's a completely custom field invented by the user.

Now that has been said, the weakness you're pointing at is a real limit of using the library through `Track` exclusively : if you use `Track` to add an additional field on a file that has multiple tagging formats, that field's key will be added _as is_ on all formats, regardless of it falling into that format's standard or not.

If you want to get surgical and handle tagging format specifics, you'd better use lower level features such as calling `MetaDataIOFactory` to instanciate a `IMetaDataIO` (see sample code in the unit tests).

> What I would have understood is [...]

Yep, that would have worked but would have made `Track` more difficult to understand for the majority of users who just want plain, no-brainer features to tag files. My position is that, when you understand where the limits of using `Track` are, you're ready to use lower level classes.

**Comment by sandreas** [comment]
Thank you for this quick and huge clarification... I'll try to experiment with `MetaDataIOFactory` and its implementation in `Track` to understand everything better. Seems that you have clear definitions of `where` and `what` and it seems that it is mostly well thought trough. Will take some time for me though :-)

**Comment by Zeugma440** [comment]
The implementation to your initial need has been published in today's v4.04
